llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp

```diff
@@ -2572,6 +2572,9 @@ SDValue DAGTypeLegalizer::PromoteIntOp_FRAMERETURNADDR(SDNode *N) {
 }
 
 SDValue DAGTypeLegalizer::PromoteIntOp_ExpOp(SDNode *N) {
+  if (N->getValueType(0).isVector())
+    return DAG.UnrollVectorOp(N);
```

Comment:
This probably shouldn't be the first thing tried. If a vector libcall happens to be available, that would be preferable. Can you move this down before the scalar call is introduced?

Reply:
Thanks, I will move it into this `if` statement to ensure that it won't promote the second operand. Do you think it's reasonable?
```
  if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
    if (N->getValueType(0).isVector())
      return DAG.UnrollVectorOp(N);
    SmallVector<SDValue, 3> NewOps(N->ops());
    NewOps[1 + OpOffset] = SExtPromotedInteger(N->getOperand(1 + OpOffset));
    return SDValue(DAG.UpdateNodeOperands(N, NewOps), 0);
  }
```